src/zjs_web_sockets.c

```diff
@@ -0,0 +1,793 @@
+        memcpy(out + mask_offset, payload, len);
+        return mask_offset + len;
+    }
+}
+
+// decode WS TCP data into a ws_packet_t type
+static int decode_packet(ws_packet_t *packet, uint8_t *data, uint32_t len)
+{
+    uint32_t mask_offset = 2;
+    uint8_t byte1 = data[0];
+    uint8_t byte2 = data[1];
+    packet->fin = (byte1 & 0x1);            // FIN      bytes: 0 bits: 0
+    packet->rsv1 = (byte1 & 0x2);           // RSV1     bytes: 0 bits: 1
+    packet->rsv2 = (byte1 & 0x4);           // RSV2     bytes: 0 bits: 2
+    packet->rsv3 = (byte1 & 0x8);           // RSV3     bytes: 0 bits: 3
+    packet->opcode = (byte1 & 0xf);         // OPCODE   bytes: 0 bits: 4-7
+    packet->mask_bit = (byte2 & 0x1);       // MASK_BIT bytes: 1 bits: 0
+    packet->payload_len = (byte2 & 0x7f);   // PAY LEN  bytes: 1 bits: 1-7
+    // payload length is less than 125 bytes
+    if (packet->payload_len <= 125) {
+        packet->payload_offset = 6;
+    } else if (packet->payload_len == 126) {
+        // payload length is greater than 125 bytes
+        uint8_t byte3 = data[2];            // PAY LEN  bytes: 2-3  bits: all
+        uint8_t byte4 = data[3];
+        // decode payload length
+        packet->payload_len = (byte3 << 8) | byte4;
+        packet->payload_offset = 8;
+        mask_offset = 4;
+    } else if (packet->payload_len == 127) {
+        /*
+         * TODO: is this something we should support?
+         */
+        ERR_PRINT("only support payload <= 65,535 bytes\n");
+        return -1;
+    }
+    // payload mask, 4 bytes
+    packet->mask[0] = data[mask_offset];
+    packet->mask[1] = data[mask_offset + 1];
+    packet->mask[2] = data[mask_offset + 2];
+    packet->mask[3] = data[mask_offset + 3];
+    packet->payload = zjs_malloc(packet->payload_len + 1);
+    memset(packet->payload, 0, packet->payload_len);
+    int j = 0;
+    int i;
+    // decode payload using payload mask
+    for (i = packet->payload_offset; i < packet->payload_len + packet->payload_offset; i++) {
+        packet->payload[j] = data[i] ^ packet->mask[j % 4];
+        j++;
+    }
+    packet->payload[packet->payload_len] = '\0';
+    DBG_PRINT("packet decoded: %s\n", packet->payload);
+    return 0;
+}
+
+static void trigger_data(void *h, const void *args)
+{
+    ws_connection_t *con = (ws_connection_t *)h;
+    uint32_t len_encode = *((uint32_t*)args);
+    // decode length and opcode from void* handle
+    uint16_t len = (len_encode & 0xffff);
+    uint16_t opcode = ((len_encode & 0xffff0000) >> 16);
+    zjs_buffer_t *zbuf;
+    ZVAL data_buf = zjs_buffer_create(len, &zbuf);
+    // copy data from read buffer
+    memcpy(zbuf->buffer, con->rptr, len);
+    con->rptr += len;
+    if (con->rptr == con->wptr) {
+        con->rptr = con->wptr = con->rbuf;
+        DBG_PRINT("all data consumed, reseting read buffer\n");
+    }
+    switch (opcode) {
+    case WS_PACKET_CONTINUATION:
+        // continuation frame
```

Comment:
the only difference between the two is that continuation is part of the previous packet. The data is still given to JS the same though. I have not actually encountered this case yet because there is no way to force it from the Node side.